include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -358,34 +392,20 @@ struct __get_sycl_range
 
   private:
     //We have to keep sycl buffer(s) instance here by sync reasons;
-    ::std::vector<::std::shared_ptr<oneapi::dpl::__internal::__lifetime_keeper_base>> m_buffers;
-
-    template <typename _Iter>
-    buf_type<val_t<_Iter>>
-    copy_back(_Iter __first, buf_type<val_t<_Iter>> buf, /*_copy_back*/ ::std::false_type)
-    {
-        return buf;
-    }
+    ::std::vector<::std::unique_ptr<oneapi::dpl::__internal::__lifetime_keeper_base>> m_buffers;
 
-    template <typename _Iter>
-    buf_type<val_t<_Iter>>
-    copy_back(_Iter __first, buf_type<val_t<_Iter>> buf, /*_copy_back*/ ::std::true_type)
-    {
-        buf.set_final_data(__first);
-        buf.set_write_back(true);
-        return buf;
-    }
+    static constexpr bool __is_copy_direct =
+            AccMode == sycl::access::mode::read_write || AccMode == sycl::access::mode::read;
+    static constexpr bool __is_copy_back =
+            AccMode == sycl::access::mode::read_write || AccMode == sycl::access::mode::write;
 
-    template <typename _Iter, typename _CopyDirectTag>
-    buf_type<val_t<_Iter>>
-    copy_direct(_Iter __first, _Iter __last, _CopyDirectTag)
-    {
-        //create a SYCL buffer and copy data [first, last) or create a empty SYCL buffer with size = (last - first)
-        if constexpr (_CopyDirectTag::value)
-            return sycl::buffer<val_t<_Iter>, 1>(__first, __last);
-        else
-            return sycl::buffer<val_t<_Iter>, 1>(__last - __first);
-    }
+    //SFINAE iterator type checks
+    template<typename It, typename T = decltype(std::addressof(*::std::declval<It&>()))>
+    static constexpr std::true_type __is_addressable (int);
```

Comment:
Please consider the following:

```suggestion
template<typename It>
static constexpr auto __is_addressable (int) -> decltype(std::addressof(*::std::declval<It&>()), std::true_type{});
```